v4/export/sql.go

```diff
@@ -1009,3 +1056,154 @@ func GetPartitionNames(db *sql.Conn, schema, table string) (partitions []string,
+// return (dbName -> tbName -> partitionName -> partitionID, error)
+func GetPartitionTableIDs(db *sql.Conn, tables map[string]map[string]struct{}) (map[string]map[string]map[string]int64, error) {
+	const (
+		showStatsHistogramsSQL   = "SHOW STATS_HISTOGRAMS"
+		selectStatsHistogramsSQL = "SELECT TABLE_ID,FROM_UNIXTIME(VERSION DIV 262144 DIV 1000,'%Y-%m-%d %H:%i:%s') AS UPDATE_TIME,DISTINCT_COUNT FROM mysql.stats_histograms"
+	)
+	partitionIDs := make(map[string]map[string]map[string]int64, len(tables))
+	rows, err := db.QueryContext(context.Background(), showStatsHistogramsSQL)
+	if err != nil {
+		return nil, errors.Annotatef(err, "sql: %s", showStatsHistogramsSQL)
+	}
+	results, err := GetSpecifiedColumnValuesAndClose(rows, "DB_NAME", "TABLE_NAME", "PARTITION_NAME", "UPDATE_TIME", "DISTINCT_COUNT")
+	if err != nil {
+		return nil, errors.Annotatef(err, "sql: %s", showStatsHistogramsSQL)
+	}
+	type partitionInfo struct {
+		dbName, tbName, partitionName string
+	}
+	saveMap := make(map[string]map[string]partitionInfo)
+	for _, oneRow := range results {
+		dbName, tbName, partitionName, updateTime, distinctCount := oneRow[0], oneRow[1], oneRow[2], oneRow[3], oneRow[4]
+		if len(partitionName) == 0 {
+			continue
+		}
+		if tbm, ok := tables[dbName]; ok {
+			if _, ok = tbm[tbName]; ok {
+				if _, ok = saveMap[updateTime]; !ok {
+					saveMap[updateTime] = make(map[string]partitionInfo)
+				}
+				saveMap[updateTime][distinctCount] = partitionInfo{
+					dbName:        dbName,
+					tbName:        tbName,
+					partitionName: partitionName,
+				}
+			}
+		}
+	}
+	err = simpleQuery(db, selectStatsHistogramsSQL, func(rows *sql.Rows) error {
+		var (
+			tableID                   int64
+			updateTime, distinctCount string
+		)
+		err2 := rows.Scan(&tableID, &updateTime, &distinctCount)
+		if err2 != nil {
+			return errors.Trace(err2)
+		}
+		if mpt, ok := saveMap[updateTime]; ok {
+			if partition, ok := mpt[distinctCount]; ok {
+				dbName, tbName, partitionName := partition.dbName, partition.tbName, partition.partitionName
+				if _, ok := partitionIDs[dbName]; !ok {
+					partitionIDs[dbName] = make(map[string]map[string]int64)
+				}
+				if _, ok := partitionIDs[dbName][tbName]; !ok {
+					partitionIDs[dbName][tbName] = make(map[string]int64)
+				}
+				partitionIDs[dbName][tbName][partitionName] = tableID
+			}
+		}
+		return nil
+	})
+	return partitionIDs, err
+}
+
+// GetDBInfo get model.DBInfos from database sql interface.
+// We need table_id to check whether a region belongs to this table
+func GetDBInfo(db *sql.Conn, tables map[string]map[string]struct{}) ([]*model.DBInfo, error) {
+	const tableIDSQL = "SELECT TABLE_SCHEMA,TABLE_NAME,TIDB_TABLE_ID FROM INFORMATION_SCHEMA.TABLES ORDER BY TABLE_SCHEMA"
+
+	schemas := make([]*model.DBInfo, 0, len(tables))
+	var (
+		tableSchema, tableName string
+		tidbTableID            int64
+	)
+	partitionIDs, err := GetPartitionTableIDs(db, tables)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
maybe skip when zero `partitionIDs`

Reply:
This part can't be skipped. We need tableID for non-partition tables.